gateway/ln-gateway/src/lib.rs

```diff
@@ -1612,6 +1616,75 @@ impl Gateway {
         Ok(())
     }
 
+    /// Queries the client log for payment events and returns to the user.
+    pub async fn handle_payment_log_msg(
+        &self,
+        PaymentLogPayload {
+            position,
+            limit,
+            federation_id,
+            event_kinds,
+        }: PaymentLogPayload,
+    ) -> PaymentLogResponse {
+        let federation_manager = self.federation_manager.read().await;
+        let configs = {
+            let mut dbtx = self.gateway_db.begin_transaction_nc().await;
+            dbtx.load_federation_configs().await
+        };
+
+        let all_clients = configs
+            .into_keys()
+            .map(|fed_id| {
+                (
+                    fed_id,
+                    federation_manager
+                        .client(&fed_id)
+                        .expect("invalid federation")
+                        .value(),
+                )
+            })
+            .collect::<BTreeMap<_, _>>();
+
+        let clients = match federation_id {
+            Some(id) if all_clients.contains_key(&id) => {
+                let client = *all_clients.get(&id).expect("already checked");
+                vec![(id, client)]
+            }
+            Some(_) => {
+                return PaymentLogResponse(vec![]);
+            }
+            None => all_clients.into_iter().collect::<Vec<_>>(),
+        };
+
+        let event_kinds = if event_kinds.is_empty() {
+            vec![
+                OutgoingPaymentSuccess::KIND,
+                IncomingPaymentSuccess::KIND,
+                CompleteLightningPaymentSuccess::KIND,
+            ]
+        } else {
+            event_kinds
+        };
+
+        let payment_log = futures::stream::iter(clients.clone())
+            .then(|(fed_id, client)| async move {
+                client
+                    .get_event_log(position, limit)
+                    .await
+                    .into_iter()
+                    .map(|e| (e.0, e.1, e.2, e.3, fed_id, e.4))
+                    .collect::<Vec<_>>()
+            })
+            .collect::<Vec<_>>()
+            .await
+            .into_iter()
+            .flatten()
+            .filter(|e| event_kinds.contains(&e.1))
+            .collect::<Vec<_>>();
```

Comment:
nicely written!

Reply:
Could flatten and filter be done on a stream to avoid one allocation?